lwt/happy_eyeballs_lwt.ml

```diff
@@ -1,20 +1,26 @@
-
 (* Lwt tasks are spawned:
  - create starts an asynchronous timer task
  - the actions resulting from timer are scheduled in one separate task
  - the actions returned from Happy_eyeballs.connect/event are scheduled in
    respective separate tasks
 *)
 
-(* TODO - cancellation of connection attempts *)
-
 let src = Logs.Src.create "happy-eyeballs.lwt" ~doc:"Happy Eyeballs Lwt"
 module Log = (val Logs.src_log src : Logs.LOG)
 
 let now = Mtime_clock.elapsed_ns
 
+module Name_id = Map.Make(struct
+    type t = [`host] Domain_name.t * int
+    let compare (h, id) (h', id') =
+      match Domain_name.compare h h' with
+      | 0 -> Int.compare id id'
+      | x -> x
+  end)
```

Comment:
this may be unsafe, and may need further information in the key, namely ip and port. the reason is that a connection is usually done to a remote host - and once DNS resolution finished, we'll try to connect to all ip addresses (one by one) for the given id -- id is to distinguish several clients that all want to connect to the same host (but need a distinct flow / distinct connection each).

so I thought maybe (host, id) is sufficient since there aren't any simultaneous connection attempts -- but I may be wrong (need to re-read the code ;)

Reply:
I think the `id` alone might be enough. We have a global counter that we increment for the id. On 64 bit I think that should be unique. I need to read the code some more. I am also starting to think we should maybe make the id type opaque, and explain it in the documentation - it's there in `connect` without any mention. Let's do that in another PR though.

https://github.com/roburio/happy-eyeballs/blob/94732ceeec143757e556675e07d1dc3481b3792c/src/happy_eyeballs.ml#L426-L441

Reply:
> I think the `id` alone might be enough. We have a global counter that we increment for the id. On 64 bit I think that should be unique.

Yes, that is a great suggestion to make it abstract.

Though, for each `id` (that `Waiter_map` generates) there may be multiple connection attempts, namely to each ip * port tuple one... The `id` is only used to track API clients -- there may be distinct calls who all want to reach for mirageos.org, and we'll do DNS resolution only once, but establish multiple TCP flows to the remote.

So, your analysis and mine together (there are no simultaneous TCP connection attempts for the same client) show that we indeed can use id only.

Reply:
@reynir I was too impatient, and pushed [81066ce](https://github.com/roburio/happy-eyeballs/pull/30/commits/81066ce067bc95bf17107fbeb96523cd186befcb) which makes the `id` abstract and uses `Waiter_map` for the connections :)